packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -0,0 +1,239 @@
+  /** Closes the menu. */
+  close(): void,
+  value: string,
+  setValue: (value: string) => void,
+  toggle: (strategy: FocusStrategy | null, force?: boolean) => void
+}
+
+interface ComboBoxProps<T> extends CollectionBase<T>, SingleSelection {
+  isOpen?: boolean,
+  defaultOpen?: boolean,
+  onOpenChange?: (isOpen: boolean) => void,
+  inputValue?: string,
+  defaultInputValue?: string,
+  onInputChange?: (value: string) => void,
+  onFilter?: (value: string) => void,
+  allowsCustomValue?: boolean,
+  onCustomValue?: (value: string) => void,
+  completionMode?: 'suggest' | 'complete',
+  menuTrigger?: 'focus' | 'input' | 'manual',
+  isFocused: boolean
+}
+
+function* filter<T>(nodes: Iterable<Node<T>>, filterFn: (node: Node<T>) => boolean) {
+  for (let node of nodes) {
+    if (filterFn(node)) {
+      yield node;
+    }
+  }
+}
+
+class FilteredCollection<T> implements Collection<Node<T>> {
+  private keyMap: Map<Key, Node<T>> = new Map();
+  private iterable: Iterable<Node<T>>;
+  private firstKey: Key;
+  private lastKey: Key;
+  private filterFn: (node: Node<T>) => boolean;
+
+  constructor(nodes: Iterable<Node<T>>, filterFn: (node: Node<T>) => boolean) {
+    this.filterFn = filterFn;
+    this.iterable = nodes;
+
+    let visit = (node: Node<T>) => {
+      this.keyMap.set(node.key, node);
+
+      if (node.childNodes && (node.type === 'section' || node.isExpanded)) {
+        for (let child of node.childNodes) {
+          visit(child);
+        }
+      }
+    };
+
+    for (let node of this) {
+      visit(node);
+    }
+
+    let last: Node<T>;
+    let index = 0;
+    for (let [key, node] of this.keyMap) {
+      if (last) {
+        last.nextKey = key;
+        node.prevKey = last.key;
+      } else {
+        this.firstKey = key;
+      }
+
+      if (node.type === 'item') {
+        node.index = index++;
+      }
+
+      last = node;
+    }
+
+    this.lastKey = last ? last.key : ''; // what to do in empty collection??
+  }
+
+  *[Symbol.iterator]() {
+    yield* filter(this.iterable, this.filterFn);
+  }
+
+  get size() {
+    return this.keyMap.size;
+  }
+
+  getKeys() {
+    return this.keyMap.keys();
+  }
+
+  getKeyBefore(key: Key) {
+    let node = this.keyMap.get(key);
+    return node ? node.prevKey : null;
+  }
+
+  getKeyAfter(key: Key) {
+    let node = this.keyMap.get(key);
+    return node ? node.nextKey : null;
+  }
+
+  getFirstKey() {
+    return this.firstKey;
+  }
+
+  getLastKey() {
+    return this.lastKey;
+  }
+
+  getItem(key: Key) {
+    return this.keyMap.get(key);
+  }
+}
+
+let whitespace = /\s/;
+
+export function useComboBoxState<T>(props: ComboBoxProps<T>): ComboBoxState<T> {
+  let itemsControlled = !!props.onFilter;
+  let collator = useCollator({sensitivity: 'base'});
+  /*
+  let menuControlled = props.isOpen !== undefined;
+  let valueControlled = props.inputValue !== undefined;
+  let selectedControlled = !!props.selectedKey;
+   */
+
+  let selectState = useSelectState(props);
```

Comment:
Issue arises when deselecting items on click (https://jira.corp.adobe.com/browse/RSP-1709). 
I figure combobox shouldn't even support deselecting items (except through clearing the input field's value) so maybe I'll add a option to useSelectableItem that disables calling toggle onselect